Replace all remaining browser dialogs with modals #5734

YongJieYongJie opened the pull request:
Fixes #5734

<!--
Please follow the naming conventions in the "guidelines for contributing" link above or the pull request may be rejected.

Also, do consider outlining the solution taken as doing so will likely help in the reviewing process
(e.g., when the pull request involves non-trivial changes)
-->

YongJieYongJie commented on src/main/webapp/js/instructorHome.js:
```diff
@@ -32,6 +32,22 @@ $(document).ready(function() {
+
+        BootboxWrapper.showModalConfirmation('Confirm archiving course', messageText, okCallback, null,
+                BootboxWrapper.DEFAULT_OK_TEXT, BootboxWrapper.DEFAULT_CANCEL_TEXT, StatusType.INFO);
+
```
extra space

YongJieYongJie commented on src/main/java/teammates/ui/controller/InstructorCourseEditPageData.java:
```diff
@@ -166,51 +137,108 @@ public int getInstructorToShowIndex() {
+        String buttonContent = "<span class=\"glyphicon glyphicon-pencil\"></span> Edit";
+        String buttonId = "courseEditLink";
+        return createBasicButton(buttonContent, buttonId, "javascript:;", Const.Tooltips.COURSE_INFO_EDIT,
+                                  isDisabled);
```
indentation

YongJieYongJie commented on src/main/webapp/js/instructor.js:
```diff
@@ -340,13 +327,25 @@ function bindRemindButtons() {
+
+        BootboxWrapper.showModalConfirmation('Confirm publishing responses', messageText, okCallback, null,
+                BootboxWrapper.DEFAULT_OK_TEXT, BootboxWrapper.DEFAULT_CANCEL_TEXT, StatusType.WARNING);
 
```
extra line

YongJieYongJie commented on src/test/java/teammates/test/cases/ui/browsertests/InstructorHomePageUiTest.java:
```diff
@@ -234,10 +234,12 @@ public void testRemindActions() {
-        homePage.clickAndConfirm(homePage.getRemindLink(feedbackSessionOpen.getCourseId(),
-                                                        feedbackSessionOpen.getFeedbackSessionName()));
+        homePage.getRemindLink(
+                feedbackSessionOpen.getCourseId(), feedbackSessionOpen.getFeedbackSessionName()).click();
```
maybe put only `.click()` on second line

wkurniawan07 commented:
Similarly here, all instances of `element.click()` should be migrated to `click(element)`.
Also, can the ESLint rule `no-alert` be enforced now?

YongJieYongJie commented:
@wkurniawan07 Yup, no-alert should be enforceable. I'll fix the erroneous merge too

wkurniawan07 commented:
@YongJieYongJie is this ready for review?

YongJieYongJie commented:
@wkurniawan07 Ready for review!

wkurniawan07 commented on src/main/java/teammates/ui/controller/InstructorCourseEditPageData.java:
```diff
@@ -69,24 +69,16 @@ public InstructorCourseEditPageData(AccountAttributes account, CourseAttributes
+        editCourseButton = createEditCourseButton(isEditDeleteButtonDisabled);
+        deleteCourseButton = createDeleteCourseButton(isEditDeleteButtonDisabled);
+
+        boolean isAddButtonDisabled = !currentInstructor.isAllowedForPrivilege(
```
`isAddInstructorButtonDisabled` because `isAddButtonDisabled` doesn't tell what object to add.

wkurniawan07 commented on src/main/java/teammates/ui/controller/InstructorCourseEditPageData.java:
```diff
@@ -69,24 +69,16 @@ public InstructorCourseEditPageData(AccountAttributes account, CourseAttributes
-    private void createButtons(InstructorAttributes currentInstructor) {
-        boolean isDisabled = !currentInstructor.isAllowedForPrivilege(
+    private void createCourseRelatedButtons(InstructorAttributes currentInstructor) {
+        boolean isEditDeleteButtonDisabled = !currentInstructor.isAllowedForPrivilege(
```
Similarly here `isEditDeleteCourseButtonDisabled`.

wkurniawan07 commented on src/main/java/teammates/ui/controller/InstructorCourseEditPageData.java:
```diff
@@ -166,38 +137,109 @@ public int getInstructorToShowIndex() {
+    }
+
+    private ElementTag createDeleteCourseButton(boolean isDisabled) {
+        String buttonContent = "<span class=\"glyphicon glyphicon-trash\"></span>Delete";
```
Although it may not have difference in the UI, let's add space between the closing span and "Delete".

wkurniawan07 commented on src/test/java/teammates/test/cases/ui/browsertests/InstructorCoursesPageUiTest.java:
```diff
@@ -291,11 +291,13 @@ public void testDeleteAction() throws Exception {
     
         String courseId = "CCAddUiTest.course1";
-        coursesPage.clickAndCancel(coursesPage.getDeleteLink(courseId));
+        coursesPage.getDeleteLink(courseId).click();
```
Need to be `click(...)` instead of `.click()`.

wkurniawan07 commented on src/test/java/teammates/test/cases/ui/browsertests/InstructorCoursesPageUiTest.java:
```diff
@@ -291,11 +291,13 @@ public void testDeleteAction() throws Exception {
 
-        coursesPage.clickAndConfirm(coursesPage.getDeleteLink(courseId))
-            .verifyHtmlMainContent("/instructorCoursesDeleteSuccessful.html");
+        coursesPage.getDeleteLink(courseId).click();
```
Similarly here.

wkurniawan07 commented on src/test/java/teammates/test/pageobjects/AppPage.java:
```diff
@@ -986,6 +1009,11 @@ private void respondToAlertWithRetry(WebElement elementToClick, boolean isConfir
         }
     }
     
+    private void clickHiddenElement(String elementId) {
```
This function looks the same as `click(...)`.

wkurniawan07 commented on src/test/java/teammates/test/pageobjects/AppPage.java:
```diff
@@ -637,6 +637,18 @@ public AppPage clickAndConfirm(WebElement elementToClick) {
+     * Fails if there is no dialog box.
+     * @return the resulting page.
+     */
+    public AppPage clickHiddenElementAndConfirm(String elementId) {
```
I may be wrong, but I think this function can replace and decommission `clickAndConfirm` (similarly with `clickAndCancel`).

wkurniawan07 commented:
Added some comments, in particular `element.click()` -> `click(element)`.
Also, don't forget to turn on the `no-alert` rule in `teammates-eslint.yml`.

YongJieYongJie commented:
@wkurniawan07 I removed the `clickHiddenElementAndConfirm()` and kept `clickAndConfirm()` instead (making the necessary modifications needed to handle modals).

I also undid previous edits where I replaced occurrences of `clickAndConfirm()` with a `click()` followed by a `waitForConfirmationModalAndClickOk()`

wkurniawan07 commented:
Looks good, for a more comprehensive review (and perhaps for formality purpose) can you attach the screenshots of all the migrated alerts here?

YongJieYongJie commented:
Here are the screenshots in context of the various pages:

Instructor Home Page:
![2016-07-29-085045_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236236/3e5e4dea-557a-11e6-9ea9-d634d7e71c31.png)
![2016-07-29-085050_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236239/3f49b17c-557a-11e6-8a1d-8c8b221350ea.png)
![2016-07-29-085058_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236241/40a13bd0-557a-11e6-9449-0a502da71861.png)
![2016-07-29-085105_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236245/4388c606-557a-11e6-8eca-8cf3ad4b2c28.png)
![2016-07-29-085110_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236246/444b63e6-557a-11e6-8fa0-a7a3d0d3c6cc.png)

YongJieYongJie commented:
Course Details Page:
![2016-07-29-085144_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236248/45b9c43e-557a-11e6-820b-f2186d6ab81f.png)
![2016-07-29-085150_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236252/4ea36e4c-557a-11e6-889c-5dc991dfc891.png)

YongJieYongJie commented:
Course Edit Page:
![2016-07-29-085631_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236257/51d45608-557a-11e6-90fd-1e1b8b9177e5.png)
![2016-07-29-085656_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236258/535a0f68-557a-11e6-8312-f29962507426.png)
![2016-07-29-085946_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236259/54f710fa-557a-11e6-8dc7-5694f168083d.png)
![2016-07-29-085955_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236260/562afc3e-557a-11e6-904d-91e4aec5fa69.png)

YongJieYongJie commented:
Other miscellaneous pages

![2016-07-29-085204_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236254/4f8b0d4c-557a-11e6-93ef-d916d0704a5c.png)
![2016-07-29-085257_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236255/509f4040-557a-11e6-9568-4eceafdbd9c2.png)
![2016-07-29-090054_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236264/597440c6-557a-11e6-8f30-49feaad6b6e8.png)
![2016-07-29-090332_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236266/5a6f412e-557a-11e6-97de-870ed08a5ecb.png)
![2016-07-29-090337_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236267/5b8e9b86-557a-11e6-984b-4682d33657fd.png)
![2016-07-29-090943_1280x800_scrot](https://cloud.githubusercontent.com/assets/14101781/17236268/5cc3e72c-557a-11e6-9afe-2b4129466d52.png)

wkurniawan07 commented:
Looks good. There may be some standardizations needed (e.g feedback session names wrapped in quotes, important names bolded) but we can leave it for another time.

wkurniawan07 commented on src/test/resources/pages/instructorCourseEditAddInstructor.html:
```diff
@@ -80,7 +80,7 @@
           <input name="user" type="hidden" value="InsCrsEdit.test">
         </form>
-        <a class="btn btn-primary btn-xs" data-original-title="Edit instructor details" data-placement="top" data-toggle="tooltip" href="javascript:;" id="instrEditLink1" title="" type="button">
+        <a class="btn btn-primary btn-xs" data-original-title="Edit course name" data-placement="top" data-toggle="tooltip" href="javascript:;" id="instrEditLink1" title="" type="button">
```
Looks like I missed this before. I don't think this change should have occurred?

wkurniawan07 commented on static-analysis/teammates-eslint.yml:
```diff
@@ -139,7 +139,7 @@ rules:
 
     # disallow the use of alert, confirm, and prompt
-    no-alert: "off" # recommended by Google
+    no-alert: 2 # recommended by Google
```
To standardize, let's make it `"error"` instead of `2`, and remove the "recommended by Google" comment.
